### PR TITLE
feat(eng-2261): License policy CLI

### PR DESCRIPTION
**Pull request by Leeoc**
Adds functionality to manage license policies via CLI

image

### CODE REVIEW

**Comment by chrisimcevoy** [comment]
Looks good at first glance. I think I confused you earlier with the whole license/licence discussion. We should adhere to "International English" in anything user-facing like commands and output. Maybe it would be better to switch back to using "license" everywhere to avoid confusion, and if you have any cases where a variable named "license" shadows the builtin, then you could just append an underscore to the variable name i.e. license_

**Comment by Leeoc** [comment]
> Looks good at first glance. I think I confused you earlier with the whole license/licence discussion. We should adhere to "International English" in anything user-facing like commands and output. Maybe it would be better to switch back to using "license" everywhere to avoid confusion, and if you have any cases where a variable named "license" shadows the builtin, then you could just append an underscore to the variable name i.e. license_

The only place `license` was an issue was for creating the base AliasGroup, just realised I can set the group to `license` and have the function def as `licence` to avoid the builtin... :noob:

File: cloudsmith_cli/cli/commands/policy/license.py [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+    if spdx_identifiers is None:
+        raise click.BadParameter(
+            "SPDX Identifiers is a required field for creating a license policy.",
+            param="spdx_identifiers",
+        )
+
+    if len(spdx_identifiers) > len(set(spdx_identifiers)):
+        raise click.BadParameter(
+            "SPDX Identifiers must be unique.", param="spdx_identifiers"
+        )
```
Comment (chrisimcevoy) [inline cloudsmith_cli/cli/commands/policy/license.py:169]: This isn't the case anymore iirc. The server will silently de-dupe them.

File: cloudsmith_cli/cli/commands/policy/license.py [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+        raise click.BadParameter(
+            "Name is a required field for creating a license policy.", param="name"
+        )
+
+    spdx_identifiers = policy_config.get("spdx_identifiers", None)
+    if spdx_identifiers is None:
+        raise click.BadParameter(
+            "SPDX Identifiers is a required field for creating a license policy.",
+            param="spdx_identifiers",
+        )
```
Comment (chrisimcevoy) [inline cloudsmith_cli/cli/commands/policy/license.py:164]: NAB: Would it be better UX to silently assign an empty list? Genuine question - I could go either way on that myself.

File: cloudsmith_cli/cli/commands/policy/license.py [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+            "owner": click.style(owner, bold=True),
+        },
+        nl=False,
+        err=use_stderr,
+    )
+
+    spdx_identifiers = policy_config.get("spdx_identifiers", [])
+    if len(spdx_identifiers) > len(set(spdx_identifiers)):
+        raise click.BadParameter(
+            "SPDX Identifiers must be unique.", param="spdx_identifiers"
```
Comment (chrisimcevoy) [inline cloudsmith_cli/cli/commands/policy/license.py:252]: Not anymore.

File: cloudsmith_cli/cli/tests/commands/policy/test_licence.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+
+    data = {
+        "name": name,
+        "description": description,
+        "spdx_identifiers": list(spdx_identifiers),
+        "allow_unknown_licenses": allow_unknown_licenses,
+        "on_violation_quarantine": on_violation_quarantine,
+    }
+
+    file_path = directory / "License-POLICY-CONFIG.json"
```
Comment (chrisimcevoy) [inline cloudsmith_cli/cli/tests/commands/policy/test_licence.py:29]: NAB: Could you make the casing consistent on the filename? Doesn't make a bit of difference, but... I'll know it's there.

File: cloudsmith_cli/cli/tests/commands/policy/test_licence.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+    assert "Getting license policies ... OK" in result.output
+    # Ignoring the exact number of results and pluralization of the word "policy"
+    # here, because we may be in a system with pre-existing policies.
+    assert "Results:" in result.output
+    assert_output_matches_policy_config(result.output, policy_config_file_path)
+
+    # Change the values in the config file
+    policy_config_file_path = create_license_policy_config_file(
+        directory=tmp_path,
+        name=policy_name,
```
Comment (chrisimcevoy) [inline cloudsmith_cli/cli/tests/commands/policy/test_licence.py:136]: If `policy_name` isn't used in the rest of this test, change this to `name=random_str()`

**Review by chrisimcevoy — APPROVED** [review]
Looks great, just a handful of tiny nitpicks. Approved in advance.

File: cloudsmith_cli/cli/commands/policy/license.py (hunk 1 above)
Comment (Leeoc) [reply cloudsmith_cli/cli/commands/policy/license.py:169]: TIL, will remove it now!

File: cloudsmith_cli/cli/commands/policy/license.py (hunk 2 above)
Comment (Leeoc) [reply cloudsmith_cli/cli/commands/policy/license.py:164]: They are required for the API, feel like an empty list defeats the point of creating the policy

Comment (chrisimcevoy) [reply cloudsmith_cli/cli/commands/policy/license.py:164]: As in `spdx_identifiers = policy_config_file.get("spdx_identifiers", [])`. Maybe?

Comment (chrisimcevoy) [reply cloudsmith_cli/cli/commands/policy/license.py:164]: Ah, but you may want to create a policy which only flags unknown licenses, no...? Anyway, not worth thinking about really. It's not that hard to specify an empty list in the config file and most people are probably going to follow the example config which includes that property.

Comment (Leeoc) [reply cloudsmith_cli/cli/commands/policy/license.py:164]: Ahh thats a good point, will default it to an empty list cheers!

**Pull request merged**